src/test/ui/consts/consts-in-patterns.rs

```diff
@@ -20,8 +20,6 @@ pub fn main() {
     assert_eq!(y, 2);
     let z = match &() {
         ZST => 9,
-        // FIXME: this should not be required
-        _ => 42,
```

Comment:
Sigh, this should be required, we have an [RFC](https://github.com/rust-lang/rfcs/pull/1445) about this.

@nikomatsakis implemented the `#[structural_match]` part of the RFC, but never got to the "const patterns are opaque for exhaustiveness checking" part, so https://github.com/rust-lang/rust/issues/31434 is still open.

Reply:
uh... I didn't know about this RFC, but as the RFC states, this is not backwards compatible, and we already break this rule for constants of struct and enum type in arbitrary depth... I don't know how we could ever do this breaking change, or even how to start linting about it.

Reply:
anyway, I'll read up on the RFC and issue and open a discussion on zulip or sth. I guess this PR is blocked until then.

Reply:
It's probably just not plausible at this juncture, and maybe never was. Honestly at this point I'd really like to settle on "some answer" regarding the semantics of constants in match patterns and not try to maintain the "structural eq limbo" we've been living in. It'd be good to collect some notes on the current status and maybe have a lang team design meeting to walk over them?

Reply:
looks to me like the PR no longer changes this test, and so it seems like this question is ... if not resolved, then at least postponed for the time being? Can I mark this as "resolved"?

Reply:
Note that

```rust
    const FOO: () = ();
    match () {
        FOO => {},
        _ => {},
    }
```

complains about the `_` arm on stable and if you remove it it compiles fine. The only change this PR does (and still does) is to give the same feature to constants of reference type.

Reply:
IOW, on top of `str` and `[u8]` we also do exhaustiveness checking for constants of type `()`? What about integers?

Cc https://github.com/rust-lang/const-eval/issues/42 -- Currently I'd prefer not to expand the set of types where consts are subject to exhaustiveness checking, *in particular* not when references are involved.

Reply:
integers do exhaustiveness checks. so do structs without private fields and enums.

Reply:
@nikomatsakis 
> Honestly at this point I'd really like to settle on "some answer" regarding the semantics of constants in match patterns and not try to maintain the "structural eq limbo" we've been living in. It'd be good to collect some notes on the current status and maybe have a lang team design meeting to walk over them?

I cannot say much about the current status (the code is hard to follow and scarcely documented) but I have given some thought recently to what structural equality *means* and how it affects pattern matching. See https://github.com/rust-lang/rust/issues/74446 and [this hackmd](https://hackmd.io/Qvrj_eOFTkCHZrhJ7f1ItA#Patterns).